File: rust/src/lib/hotkey.ts
export const IS_MAC =
  typeof navigator !== "undefined" && /mac|iphone|ipad|ipod/i.test(navigator.platform);

const SYMBOLS_MAC: Record<string, string> = {
  cmd: "⌘", command: "⌘", rcmd: "⌘", rcommand: "⌘",
  shift: "⇧", rshift: "⇧",
  alt: "⌥", option: "⌥", ralt: "⌥", roption: "⌥",
  ctrl: "⌃", control: "⌃", rctrl: "⌃",
  space: "Space",
  enter: "Enter", return: "Enter",
  tab: "Tab",
  esc: "Esc", escape: "Esc",
  delete: "Delete",
  backspace: "Backspace",
};

const SYMBOLS_WIN: Record<string, string> = {
  cmd: "Ctrl", command: "Ctrl", rcmd: "Ctrl", rcommand: "Ctrl",
  ctrl: "Ctrl", control: "Ctrl", rctrl: "Ctrl",
  alt: "Alt", option: "Alt", ralt: "Alt", roption: "Alt",
  shift: "Shift", rshift: "Shift",
  win: "Win", super: "Win", meta: "Win",
  space: "Space",
  enter: "Enter", return: "Enter",
  tab: "Tab",
  esc: "Esc", escape: "Esc",
  delete: "Delete",
  backspace: "Backspace",
};

const SYMBOLS: Record<string, string> = IS_MAC ? SYMBOLS_MAC : SYMBOLS_WIN;

export function parseHotkeyKeys(combo: string): string[] {
  return combo
    .split("+")
    .map((k) => k.trim().toLowerCase())
    .filter(Boolean);
}

export function formatHotkeyKey(key: string): string {
  return SYMBOLS[key] ?? key.toUpperCase();
}

/// 把 "cmd+shift+f5" 格式化成 "⌘⇧F5"，用在文本里（非 <kbd>）
export function formatHotkey(combo: string): string {
  return parseHotkeyKeys(combo).map(formatHotkeyKey).join("");
}

// 所有 modifier 的小写名字；和 Rust hotkey.rs 里认的保持一致
const MODIFIERS = new Set([
  "cmd", "command", "rcmd", "rcommand",
  "ctrl", "control", "rctrl",
  "alt", "option", "ralt", "roption",
  "shift", "rshift",
  "win", "super", "meta",
]);

/// 把 KeyboardEvent.code 转成 Rust hotkey.rs 认识的主键字符串：
/// KeyA → "a"、Digit1 → "1"、F5 → "f5"、Space → "space"、Enter → "enter" 等。
/// 比直接用 e.key 稳定——不受 shift 大小写、输入法、AltGr 影响。
export function keyCodeToName(code: string, fallback: string): string {
  if (code.startsWith("Key")) return code.slice(3).toLowerCase();
  if (code.startsWith("Digit")) return code.slice(5);
  if (/^F\d{1,2}$/.test(code)) return code.toLowerCase();
  const SPECIAL: Record<string, string> = {
    Space: "space",
    Enter: "enter",
    Tab: "tab",
    Backspace: "backspace",
    Delete: "delete",
    Escape: "esc",
  };
  return SPECIAL[code] ?? fallback.toLowerCase();
}

/// 前端实时校验，和 Rust hotkey::to_tauri_shortcut 的核心约束对齐：
/// - 至少有 2 段（mod+key）
/// - 至少有一个非 modifier 的主键
/// 返回 null 表示通过，否则返回人话错误提示。
export function validateHotkey(combo: string): string | null {
  const keys = parseHotkeyKeys(combo);
  if (keys.length === 0) return "快捷键不能为空";
  if (keys.length < 2) return "至少要有一个修饰键（Cmd/Ctrl/Shift/Alt）+ 一个主键";
  const nonMods = keys.filter((k) => !MODIFIERS.has(k));
  if (nonMods.length === 0) {
    return "缺少主键：需要 F1–F24、A–Z、0–9 或 Space/Enter/Tab/Esc 等";
  }
  return null;
}

/// 把 double_tap_modifier 字符串(如 "right_option")格式化成符号(如 "⌥右")用于
/// indicator hint 等显示。
export function formatDoubleTapModifier(value: string): string {
  const sideMap: Record<string, string> = { left: "左", right: "右" };
  const symbolMap: Record<string, string> = IS_MAC
    ? { option: "⌥", alt: "⌥", ctrl: "⌃", control: "⌃", shift: "⇧", cmd: "⌘", command: "⌘", fn: "Fn" }
    : { option: "Alt", alt: "Alt", ctrl: "Ctrl", control: "Ctrl", shift: "Shift", cmd: "Win", command: "Win", fn: "Fn" };
  if (value === "fn") return "Fn";
  const m = /^(left|right)_(\w+)$/.exec(value);
  if (!m) return value;
  const [, side, key] = m;
  const symbol = symbolMap[key] ?? key;
  // macOS 上 ⌥右 这种写法清楚;Windows 上 "右 Alt" 更直观
  return IS_MAC ? `${symbol}${sideMap[side] ?? ""}` : `${sideMap[side] ?? ""} ${symbol}`;
}

/// 触发方式总入口:Toggle/PTT 显示 hotkey,DoubleTapHold 显示"双击 ⌥右"。
/// App.tsx sidebar / HistoryView 顶 / indicator hint 都用这个。
export function formatTrigger(cfg: {
  trigger_mode?: string;
  hotkey: string;
  double_tap_modifier?: string;
}): string {
  if (cfg.trigger_mode === "double_tap_hold" && cfg.double_tap_modifier) {
    return `双击 ${formatDoubleTapModifier(cfg.double_tap_modifier)}`;
  }
  return formatHotkey(cfg.hotkey);
}

File: rust/src-tauri/src/hotkey.rs
//! 热键字符串解析：把用户配置的 "cmd+shift+f5" 转成 tauri-plugin-global-shortcut 可接受的格式。
//! 对应 Go 版的 hotkey.go。

use anyhow::{bail, Result};

/// 把用户风格的热键字符串（如 "cmd+shift+f5"、"ctrl+alt+space"）
/// 标准化成 tauri shortcut 字符串（如 "CommandOrControl+Shift+F5"）。
///
/// - cmd / command / rcmd / rcommand → CommandOrControl
/// - ctrl / rctrl → Control
/// - option / alt / ralt / roption → Alt
/// - shift / rshift → Shift
/// - win / super / meta → Super（Windows 键 / Super 键）
/// - 主键（a-z, 0-9, f1-f24, space/tab/...）保持字母并首字母大写
pub fn to_tauri_shortcut(input: &str) -> Result<String> {
    let parts: Vec<&str> = input.trim().split('+').map(str::trim).collect();
    if parts.len() < 2 {
        bail!("热键格式应为 mod+key，如 cmd+shift+f5");
    }

    let mut mods: Vec<&str> = Vec::new();
    let mut key: Option<String> = None;

    for raw in parts {
        let p = raw.to_lowercase();
        match p.as_str() {
            "cmd" | "command" | "rcmd" | "rcommand" => push_unique(&mut mods, "CommandOrControl"),
            "ctrl" | "control" | "rctrl" => push_unique(&mut mods, "Control"),
            "option" | "alt" | "ralt" | "roption" => push_unique(&mut mods, "Alt"),
            "shift" | "rshift" => push_unique(&mut mods, "Shift"),
            "win" | "super" | "meta" => push_unique(&mut mods, "Super"),
            other => {
                if other.is_empty() {
                    bail!("热键段为空");
                }
                key = Some(normalize_key(other)?);
            }
        }
    }

    let key = key.ok_or_else(|| anyhow::anyhow!("需要一个主键（如 space、a-z）"))?;
    #[cfg(target_os = "windows")]
    {
        if mods.contains(&"Super") && key == "Space" {
            bail!("win+space 是 Windows 系统保留快捷键（输入法/语言切换），无法注册");
        }
    }
    mods.push(&key);
    Ok(mods.join("+"))
}

fn push_unique(mods: &mut Vec<&str>, m: &'static str) {
    if !mods.contains(&m) {
        mods.push(m);
    }
}

fn normalize_key(p: &str) -> Result<String> {
    // 单个字母 / 数字
    if p.len() == 1 {
        let c = p.chars().next().unwrap();
        if c.is_ascii_alphanumeric() {
            return Ok(c.to_ascii_uppercase().to_string());
        }
    }
    // F 键
    if let Some(n) = p.strip_prefix('f') {
        if let Ok(num) = n.parse::<u8>() {
            if (1..=24).contains(&num) {
                return Ok(format!("F{}", num));
            }
        }
    }
    // 特殊键
    match p {
        "space" => Ok("Space".into()),
        "return" | "enter" => Ok("Enter".into()),
        "tab" => Ok("Tab".into()),
        "esc" | "escape" => Ok("Escape".into()),
        "delete" => Ok("Delete".into()),
        "backspace" => Ok("Backspace".into()),
        other => {
            bail!("未知按键: {}", other);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_basic_combo() {
        assert_eq!(
            to_tauri_shortcut("cmd+shift+f5").unwrap(),
            "CommandOrControl+Shift+F5"
        );
    }

    #[test]
    fn parses_with_option() {
        assert_eq!(
            to_tauri_shortcut("ctrl+alt+space").unwrap(),
            "Control+Alt+Space"
        );
    }

    #[test]
    fn normalizes_rmod_to_same() {
        // 修饰键顺序按用户输入保留，主键在最后
        assert_eq!(
            to_tauri_shortcut("rshift+rcmd+a").unwrap(),
            "Shift+CommandOrControl+A"
        );
        assert_eq!(
            to_tauri_shortcut("rcmd+rshift+a").unwrap(),
            "CommandOrControl+Shift+A"
        );
    }

    #[test]
    fn parses_super_combo() {
        assert_eq!(to_tauri_shortcut("win+shift+f5").unwrap(), "Super+Shift+F5");
    }

    #[test]
    fn rejects_no_key() {
        assert!(to_tauri_shortcut("cmd+shift").is_err());
    }

    #[test]
    fn rejects_empty() {
        assert!(to_tauri_shortcut("").is_err());
    }

    #[test]
    fn rejects_unknown_key() {
        assert!(to_tauri_shortcut("cmd+XYZ").is_err());
    }

    #[test]
    fn fkey_range() {
        assert_eq!(
            to_tauri_shortcut("cmd+f12").unwrap(),
            "CommandOrControl+F12"
        );
        assert!(to_tauri_shortcut("cmd+f99").is_err());
    }
}
